_drafts/device-logging.md

```diff
@@ -0,0 +1,273 @@
+
+I distinctly remember going to out to a customer’s store and connecting to their network so I could pull logs off of a remotely accessible device, just to diagnose an issue they were having.
+
+This is of course, unsustainable, if you have hardware that’s distributed across multiple locations, and if you have the want to collect logs faster than just waiting until you are physically connected to the device again.
+
+In that case, you’ll want to build a way to upload logs to a backend somewhere, account for all the storage and privacy-related concerns and regulations, and then a way to download those logs from your computer to analyze them.
+
+### Pull versus Push remote collection triggers
+
+Naively, the easiest way to upload logs is have the user do it. If the device has a UI, it’s as simple as building a button for the user to click when they’re having problems &mdash; think about the "Shake to Report Feedback" or "Submit Feedback" feature on many apps. This way, you can be upfront about the data you’re collecting right, at the time of collection.
+
+Sadly, this depends on the user though, and users are notoriously unreliable. If they’re a technical user, they might not even need an UI! — they might be happy using a CLI tool instead. But if they’re non-technical, even asking them to navigate to a certain page might be a huge burden.
+
+The first alternative that comes to mind is a push system for retrieving logs &mdash; send a notification from the server to the device, and the device will upload the logs when it receives the message. This can take the form of an actual push notification mechanism with a connected web socket, or something as simple as having the device perform a poll every once in a while and collecting the log if the poll tells it to.
+
+The problem with using a push to upload logs is that by the time you send the push, or by the time that the device receives it, the log cache might not contain the data from the time of the interest anymore. At a previous company, it was often the case that because our log cache often saved the last 2-3 hours of logs, by the time our engineers were informed about an incident in the field, a day would have passed and the logs that we’d retrieve wouldn’t have any useful information from the time of the incident.
+
+One of the cooler [log collection triggers](https://docs.memfault.com/docs/android/android-logging#log-collection) we developed here at Memfault is what I like to call "Just in time" log collection. Instead of having the user of the device click a button or someone send a push to the device, we can automatically detect if an abnormal event happened, and then mark the log files around that time as significant, where they will then be uploaded at the next convenient moment. This allows us to capture the exact log files that are relevant to an incident, if we can automatically detect that abnormal event.
```

Comment:
I feel like this docs link doesn't show off cool log collection triggers the way I'd want it to...is there somewhere else we can link to the illustrates this, or maybe we can beef up this section of the docs before publishing? Would super love to see this link wow people if they click on it!

Reply:
Haha I had this same feeling when I wrote this - let me make a quick update to the docs. I thought we had something more substantial to be honest.

Reply:
https://github.com/memfault/memfault-docs/pull/755

Reply:
Sweet!! Love it

Reply:
Going to keep this thread unresolved until you merge it so we don't forget